lib/lroPollStrategy.ts

```diff
@@ -230,21 +245,61 @@ function isFinished(status: LongRunningOperationStates): boolean {
  * @param azureServiceClient The AzureServiceClient that was used to send the initial request.
  * @param options Any options that were provided to the initial request.
  */
-export function createLROPollStrategy(initialResponse: HttpOperationResponse, azureServiceClient: AzureServiceClient, options?: RequestOptionsBase): LROPollStrategy | undefined {
-  let result: LROPollStrategy | undefined;
+export function createLROPollStrategyFromInitialResponse(initialResponse: HttpOperationResponse, azureServiceClient: AzureServiceClient, options?: RequestOptionsBase): LROPollStrategy | undefined {
+  const initialRequestMethod: HttpMethods = initialResponse.request.method;
+  const initialResponseStatus: number = initialResponse.status;
 
+  let lroPollStrategyType: LROPollStrategyType | undefined;
   if (getAzureAsyncOperationHeaderValue(initialResponse)) {
-    result = new AzureAsyncOperationLROPollStrategy(initialResponse, azureServiceClient, options);
+    lroPollStrategyType = "AzureAsyncOperation";
   } else if (getLocationHeaderValue(initialResponse)) {
-    result = new LocationLROPollStrategy(initialResponse, azureServiceClient, options);
-  } else if (initialResponse.request.method === "PUT" || initialResponse.request.method === "PATCH") {
-    result = new GetResourceLROPollStrategy(initialResponse, azureServiceClient, options);
-  } else if (initialResponse.status === 201 || initialResponse.status === 202 || isFinished(getStatusFromResponse(initialResponse))) {
-    result = undefined;
+    lroPollStrategyType = "Location";
+  } else if (initialRequestMethod === "PUT" || initialRequestMethod === "PATCH") {
+    lroPollStrategyType = "GetResource";
+  } else if (initialResponseStatus !== 201 && initialResponseStatus !== 202 && !isFinished(getStatusFromResponse(initialResponse))) {
+    throw new Error("Can't determine long running operation polling strategy.");
+  }
+
+  let result: LROPollStrategy | undefined;
+  if (lroPollStrategyType) {
+    const resource: any = getResponseBody(initialResponse);
+    const lroMemento: LROMemento = {
+      pollStrategyType: lroPollStrategyType,
+      options: options,
+      initialResponse: initialResponse,
+      mostRecentResponse: initialResponse,
+      mostRecentRequest: initialResponse.request,
+      azureAsyncOperationHeaderValue: getAzureAsyncOperationHeaderValue(initialResponse),
+      locationHeaderValue: getLocationHeaderValue(initialResponse),
+      resource: resource,
+      state: getStatusFromResponse(initialResponse, resource)
+    };
+    result = createLROPollStrategyFromMemento(azureServiceClient, lroMemento);
   } else {
-    throw new Error("Can't determine long running operation polling strategy from initial response.");
+    result = undefined;
```

Comment:
Does it ever make sense to hit this line?

Reply:
This line shouldn't ever be hit if all of our services properly follow the LRO specifications, but if they do something that is unexpected or if the initial response indicates that the operation has already completed, then this line will be hit.